### PR TITLE
Implement LSP client-side file-watcher for langserver

### LINKED ISSUE #21700: Detect the changes to .toml files in Ballerina source directory
**Description:**
Language server cannot detect and update the language-server context when the .toml files are updated.

**Steps to reproduce:**
For a Ballerina project, try to build AST with incorrect paths to the dependencies in Ballerina.toml. Then correct the dependencies, and try to build the AST again. 

**Affected Versions:**
v1.2.0-SNAPSHOT at least

### LINKED ISSUE #28125: Support for workspace events in langserver
**Description:**
Currently language server does not listens the changes of the workspace. Need to address followings;

- Creating `Kubernetes.toml`, `Dependencies.toml`, `Ballerina.toml` and `Config.toml` once the project is opened
- Changes to the project structure (such as adding/removing module)

**Affected Versions:**
2.0.0-SLP9-SNAPSHOT

### PR DESCRIPTION
**Pull request by rasika**
Fixes #28125 
Fixes #21700

## Check List 
- [ ] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/workspace/BallerinaWorkspaceManager.java [hunk 1]
```diff
@@ -256,36 +268,356 @@ public void didOpen(Path filePath, DidOpenTextDocumentParams params) throws Work
...
+                Path parent = filePath.getParent();
+                if (ProjectConstants.TEST_DIR_NAME.equals(parent.getFileName().toString())) {
+                    // If inside a tests folder, get parent
+                    parent = parent.getParent();
+                }
+                if (ProjectConstants.MODULES_ROOT.equals(parent.getParent().getFileName().toString())) {
+                    // If inside a module folder, get parent
+                    parent = parent.getParent().getParent();
+                }
+                return projectPair(parent);
```
Comment (nadeeshaan) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/workspace/BallerinaWorkspaceManager.java:383]: Any specific reason that we can't use `getProjectPair` here?

Comment (rasika) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/workspace/BallerinaWorkspaceManager.java:383]: `projectPair` is a direct map lookup. `getProjectPair` can cause creating new project. I'll send a commit to change the name `getProjectPair`

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/BallerinaWorkspaceService.java [hunk 2]
```diff
@@ -85,7 +89,23 @@ public void didChangeConfiguration(DidChangeConfigurationParams params) {
...
+            String uri = fileEvent.getUri();
+            Optional<Path> optFilePath = CommonUtil.getPathFromURI(uri);
+            if (optFilePath.isEmpty()) {
+                continue;
+            }
+            Path filePath = optFilePath.get();
+            try {
+                workspaceManager.didChangeWatched(filePath, fileEvent);
+            } catch (UserErrorException e) {
+                this.clientLogger.notifyUser("File Change Failed to Handle", e);
```
[the remark this answers is not in the document]
Comment (rasika) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/BallerinaWorkspaceService.java:103]: +1, will add

[the remark this answers is not in the document]
Comment (rasika) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/BallerinaWorkspaceService.java:103]: @nadeeshaan `UserErrorException` are meant to display as popup message boxes in right-bottom corner. Thus having a url will make it hard to read. The error message `e.getMessage()` will be displayed as body of the message. For instance in such scenario we should throw as `new UserErrorException("Couldn't handle file uri: /root/a.bal")`. BTW we haven't use any `UserErrorException` yet in the didChangeWatched flow.

[the remark this answers is not in the document]
Comment (nadeeshaan) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/BallerinaWorkspaceService.java:103]: Oh, my bad. Misread the client logger. :D.

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/workspace/BallerinaWorkspaceManager.java (hunk 1 above)
Comment (nadeeshaan) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/workspace/BallerinaWorkspaceManager.java:374]: +1

**Review by nadeeshaan — APPROVED** [review]

**Pull request merged**